src/self_driving/pilot_util.cpp

```diff
@@ -115,13 +241,31 @@ void PilotUtil::InferenceWithFeatures(
     }
     inference_result.emplace(ou_map_it.first, res.first);
   }
+
+  // populate pipeline_to_prediction using pipeline_to_ou_position and inference_result
+  std::vector<std::vector<double>> ppl_res;
+  for (auto ppl_to_ou : pipeline_to_ou_position) {
+    for (auto ou_it : std::get<2>(ppl_to_ou)) {
+      ppl_res.push_back(inference_result.at(ou_it.first)[ou_it.second]);
+    }
+    auto ppl_key = std::make_pair(std::get<0>(ppl_to_ou), std::get<1>(ppl_to_ou));
+    if (pipeline_to_prediction->find(ppl_key) == pipeline_to_prediction->end()) {
+      pipeline_to_prediction->emplace(ppl_key, std::vector<std::vector<std::vector<double>>>());
+    }
+    pipeline_to_prediction->at(ppl_key).emplace_back(std::move(ppl_res));
+  }
 }
 
 void PilotUtil::GroupFeaturesByOU(
     std::list<std::tuple<execution::query_id_t, execution::pipeline_id_t,
                          std::vector<std::pair<ExecutionOperatingUnitType, uint64_t>>>> *pipeline_to_ou_position,
+    const std::vector<execution::query_id_t> &pipeline_qids,
     const std::list<metrics::PipelineMetricRawData::PipelineData> &pipeline_data,
     std::unordered_map<ExecutionOperatingUnitType, std::vector<std::vector<double>>> *ou_to_features) {
```

Comment:
Document this function as well.

Reply:
Sure, I have put it in the header file

Reply:
Same here. I was actually referring to the code logic of the function implementation in the cpp file.